spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/CachingClientConnectionFactory.java

```diff
@@ -225,17 +225,30 @@ public String toString() {
 		 */
 		@Override
 		public boolean onMessage(Message<?> message) {
-			AbstractIntegrationMessageBuilder<?> messageBuilder =
-					CachingClientConnectionFactory.this.getMessageBuilderFactory()
-							.fromMessage(message)
-							.setHeader(IpHeaders.CONNECTION_ID, getConnectionId());
-			if (message.getHeaders().get(IpHeaders.ACTUAL_CONNECTION_ID) == null) {
-				messageBuilder.setHeader(IpHeaders.ACTUAL_CONNECTION_ID,
-						message.getHeaders().get(IpHeaders.CONNECTION_ID));
+			Message<?> modifiedMessage;
+			if (message instanceof ErrorMessage) {
+				Map<String, Object> headers = new HashMap<String, Object>(message.getHeaders());
+				headers.put(IpHeaders.CONNECTION_ID, getConnectionId());
+				if (headers.get(IpHeaders.ACTUAL_CONNECTION_ID) == null) {
+					headers.put(IpHeaders.ACTUAL_CONNECTION_ID,
+							message.getHeaders().get(IpHeaders.CONNECTION_ID));
+				}
+				modifiedMessage = new ErrorMessage((Throwable) message.getPayload(), headers);
```

Comment:
So, shouldn't our `MessageBuilder` take care about `ErrorMessage` modification to avoid such a copy/paste cases?
I don't say to do this right now, but just in case we should pay attention.